bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/regexp/TokenReader.java

```diff
@@ -0,0 +1,188 @@
+        this.treeTraverser = treeTraverser;
+    }
+
+    /**
+     * Consumes the input and return the next token.
+     *
+     * @return Next token in the input
+     */
+    public Token read() {
+        if (this.tokensAhead.size > 0) {
+            // Cache the head.
+            this.currentToken = tokensAhead.consume();
+            return this.currentToken;
+        }
+
+        // Cache the head.
+        this.currentToken = this.treeTraverser.nextToken();
+        return this.currentToken;
+    }
+
+    /**
+     * Lookahead in the input and returns the next token. This will not consume the input.
+     * That means calling this method multiple times will return the same result.
+     *
+     * @return Next token in the input
+     */
+    public Token peek() {
+        if (this.tokensAhead.size > 0) {
+            return this.tokensAhead.peek();
+        }
+
+        Token token = this.treeTraverser.nextToken();
+        this.tokensAhead.add(token);
+        return token;
+    }
+
+    /**
+     * Lookahead in the input and returns the token at the k-th position from the current
+     * position of the input token stream. This will not consume the input. That means
+     * calling this method multiple times will return the same result.
+     *
+     * @param k Position of the character to peek
+     * @return Token at the k-position from the current position
+     */
+    public Token peek(int k) {
+        Token nextToken;
+        TokenBuffer tokensAhead = this.tokensAhead;
+
+        while (tokensAhead.size < k) {
+            nextToken = this.treeTraverser.nextToken();
+            if (tokensAhead.size == tokensAhead.capacity) {
+                // We reach here when the BUFFER_SIZE is exceeded.
+                // To avoid parser being crashed, return EOF token as peek(k) for k > BUFFER_SIZE.
+                return createToken(TokenKind.EOF_TOKEN);
+            }
+
+            tokensAhead.add(nextToken);
+        }
+
+        return tokensAhead.peek(k);
+    }
+
+    public static Token createToken(TokenKind kind) {
+        return new Token(kind);
+    }
+
+    /**
+     * A ring buffer of tokens.
+     *
+     * @since 2201.3.0
+     */
+    private static class TokenBuffer {
+
+        private final int capacity;
+        private final Token[] tokens;
+        private int endIndex = -1;
+        private int startIndex = -1;
+        private int size = 0;
+
+        TokenBuffer(int size) {
+            this.capacity = size;
+            this.tokens = new Token[size];
+        }
+
+        /**
+         * Add a token to the buffer.
+         *
+         * @param token Token to add
+         */
+        public void add(Token token) {
+            if (this.size == this.capacity) {
+                throw new IndexOutOfBoundsException("buffer overflow");
+            }
+
+            if (this.endIndex == this.capacity - 1) {
+                this.endIndex = 0;
+            } else {
+                this.endIndex++;
+            }
```

Comment:
Shall we do a math operation like this
`this.endIndex = this.endIndex++ % this.capacity;`